### PR TITLE
First stab at sidebars_widgets-based widget management

### PR DESCRIPTION
**Pull request by adamziel**
## Description
This is an attempt at making the experimental widget management screen fully functional so that it's possible to mix&match legacy widgets and gutenberg blocks.

It works as follows:
* wp_options are embraced, posts-based manipulations are removed (see https://github.com/WordPress/gutenberg/issues/24267).
* Blocks are just regular widgets implemented using the `WP_Widget_Block` class.
* Frontend is based on the data layer where we use widgetArea as the entity. It has a corresponding API endpoint which returns a widget area with all its widgets and it also makes it possible to POST a widget area with a set of desired widgets.
* Legacy widgets on the new screen are no longer saved on change, in fact they're not saved at all until a "Save" button is clicked.

Other considerations:
* Consider making widget a base entity instead of widget area - batch saving is a pre-requisite for this approach though, also it would require some discussion about how reordering widgets would work.
* The endpoint says /sidebars but the entity name is widgetArea - it would be good to address this discrepancy somehow.
* Customizer will only display rendered blocks instead of an actual Gutenberg instance - let's discuss.
* `/widget-utils` controller is still there and would be a good candidate for removal in some future iteration.
* Overall this PR adds a bunch of code behind a feature flag, merging it and then iterating on bugfixes would make a sensible approach.

Reuses parts of the following plugin (CC @martin-pettersson) https://github.com/martin-pettersson/wp-rest-api-sidebars

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by noisysocks** [comment]
Looks promising. It's amazing just how similar this ends up being to how the Navigation screen works.

[wp-rest-api-sidebars](https://github.com/martin-pettersson/wp-rest-api-sidebars) is a nice find! I think let's copy `class-sidebars-controller.php` and `class-widgets-controller.php` into `lib`. That will make it easier to test this PR and see exactly what code from the plugin we're proposing to add to Core.

File: packages/core-data/src/resolvers.js
Comment (adamziel) [inline packages/core-data/src/resolvers.js:57]: seehttps://github.com/WordPress/gutenberg/issues/24308

File: lib/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-sidebars-controller.php:55]: short array syntax isn't allowed.

File: lib/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-sidebars-controller.php:58]: This will need a `permission_callback`.

File: lib/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-sidebars-controller.php:72]: Missing translation functions.

File: lib/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-sidebars-controller.php:69]: Needs a permission callback.

File: lib/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-sidebars-controller.php:89]: This method should be added to the single route `register_rest_route` call.

File: lib/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-sidebars-controller.php:59]: We need a schema.

File: lib/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-sidebars-controller.php:107]:
```suggestion
		$sidebar_id    = $request['id'];
```

Just use the request object directly.

File: lib/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-sidebars-controller.php:143]: This needs to be slashed. REST API request data is unslashed, but the PHP globals are slashed.

File: lib/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-sidebars-controller.php:176]: This can be removed.

File: lib/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-sidebars-controller.php:197]: This should be refactored to use `_fields` and/or `context`.

File: lib/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-sidebars-controller.php:216]: This can be dropped.

**Review by TimothyBJacobs — DISMISSED** [review]
I've left some preliminary REST API feedback. I think this general approach makes sense.

> /widget-utils controller is still there and would be a good candidate for removal in some future iteration.

Yeah I'd like to drop this as well.

Ideally I think the widgets would be in their own endpoint. So `sidebars/<id>/widgets` and `sidebars/<id>/widgets/<id>`.

----

Is this screen still behind a Gutenberg experiment checkbox as well? If so, ideally we would only register the REST API routes if the experiment is enabled.

----

This will also require phpunit tests for the REST API endpoint.

File: packages/core-data/src/resolvers.js
Comment (youknowriad) [reply packages/core-data/src/resolvers.js:57]: this change is problematic for me. `entity` will always be defined aside on the very first call so `getEntityRecord`  won't do anything for all calls except the first one?

File: packages/edit-widgets/src/components/widget-areas-block-editor-content/index.js
Comment (youknowriad) [inline packages/edit-widgets/src/components/widget-areas-block-editor-content/index.js:59]: A separate editor per widget area is how it was written initially but for design reasons I believe @jorgefilipecosta refactored this to be a single editor instance:

I agree with you that separate editors are conceptually easier to reason about but a single editor has some advantages too:

 - The global inserter works easily
 - You can move blocks between widget areas.

See #22140 for more details

File: packages/core-data/src/resolvers.js
Comment (adamziel) [reply packages/core-data/src/resolvers.js:57]: @youknowriad I replied in https://github.com/WordPress/gutenberg/issues/24308

Comment (adamziel) [reply packages/core-data/src/resolvers.js:57]: To close the loop here - this change should **not** be merged, it just happens to be useful on the widget screen exclusively until #19498 lands.

File: lib/class-wp-rest-sidebars-controller.php
Comment (adamziel) [reply lib/class-wp-rest-sidebars-controller.php:171]: Good spot! It seems like [class-wp-widget.php](https://github.com/WordPress/WordPress/blob/master/wp-includes/class-wp-widget.php#L433) immediately calls `stripslashes_deep` on that data so it should go in there in the slashed form.

File: packages/edit-widgets/src/components/widget-areas-block-editor-content/index.js
Comment (adamziel) [reply packages/edit-widgets/src/components/widget-areas-block-editor-content/index.js:59]: Thank you for this extra context, a single editor instance definitely has upsides - what I found hard was making the blocks live in redux state instead of `const blocks = useState()`. I will keep exploring and possibly follow up with a few questions.

Comment (adamziel) [reply packages/edit-widgets/src/components/widget-areas-block-editor-content/index.js]: I got it to work as a single editor using a similar "ephemeral post" as on the navigation management screen. I think it would also be possible to get rid of the "widget area post" on only have that top level post and widget area entities, but it's a tactical detail.

**Comment by noisysocks** [comment]
If I register a widget that doesn't use `WP_Widget` to a sidebar and then go to Gutenberg → Widgets (beta), nothing loads and I see a 500 error in DevTools.

Here's the plugin I made to test with:

```php
function marquee_greeting_init() {
	wp_register_sidebar_widget(
		'marquee_greeting',
		'Marquee Greeting',
		function() {
			$greeting = get_option( 'marquee_greeting', 'Hello!' );
			printf( '<marquee>%s</marquee>', esc_html( $greeting ) );
		}
	);

	wp_register_widget_control(
		'marquee_greeting',
		'Marquee Greeting',
		function() {
			if ( isset( $_POST['marquee-greeting'] ) ) {
				update_option(
					'marquee_greeting',
					sanitize_text_field( $_POST['marquee-greeting'] )
				);
			}

			$greeting = get_option( 'marquee_greeting' );
			?>
			<p>
				<label for="marquee-greeting">Greeting:</label>
				<input
					id="marquee-greeting"
					class="widefat"
					name="marquee-greeting"
					type="text"
					value="<?= esc_attr( $greeting ) ?>"
					placeholder="Hello!" />
			</p>
			<?php
		}
	);
}
add_action( 'init', 'marquee_greeting_init' );
```

File: packages/edit-widgets/src/store/selectors.js
Comment (adamziel) [reply packages/edit-widgets/src/store/selectors.js:31]: Eagle eye! It should have been invoked in `WidgetAreasBlockEditorProvider` but it wasn't, I just added it there.

File: packages/core-data/src/entities.js
Comment (adamziel) [reply packages/core-data/src/entities.js:64]: I'm not convinced about this part, `sidebar` is a terribly unintuitive name for a widget area. I understand it's the standard name used in core too and so I'm torn.

Comment (adamziel) [reply packages/core-data/src/entities.js:64]: I renamed just the entity and retained the `widgetArea` name in selectors, resolvers etc. Happy to discuss it further in a  follow-up issue.

File: lib/class-wp-widget-block.php
Comment (adamziel) [reply lib/class-wp-widget-block.php:83]: They are now 👍

File: packages/block-library/src/legacy-widget/edit/dom-manager.js
Comment (adamziel) [reply packages/block-library/src/legacy-widget/edit/dom-manager.js:72]: This code was autosaving the legacy widgets on each change so I removed it - there is a save button on the widget management screen so autosaving on change is at least unexpected.

**Comment by adamziel** [comment]
> If I add a widget that doesn't use WP_Widget to a sidebar and then go to Gutenberg → Widgets (beta), nothing loads and I see a 500 error in DevTools.

Thank you for a complete example, it made testing so much easier for me! This PR will now work with these "old-style" widgets.

**Comment by adamziel** [comment]
I went ahead and added PHPUnit tests for the endpoint. Also everything related to the new widgets screen is now loaded only when the experiment flag is enabled. I also removed the data layer change. CC @noisysocks

File: packages/core-data/src/entities.js
Comment (noisysocks) [reply packages/core-data/src/entities.js:64]: Agree that it's a regrettable name, but I'd personally prefer that the combined codebase (Core+Gutenberg) has one regrettable name for something instead of two names for something.

**Review by noisysocks — COMMENTED** [review]
I gave this a spin and noticed lots of bugs 😅. I think a lot of these exist in `master`, though, so I'll leave it up to you whether you want to fix any of them now or as follow-up PRs. This is the last architectural change I envision for this feature, so I'm happy to see this refactor merged quickly and the feature iterated on rapidly ahead of removing the experimental flag for WP 5.6.

Notes:

- Thanks for adding PHP unit tests!
- We should add unit tests for all of the exported function sin `packages/edit-widgets/store`.
- We should, later, add E2E tests for the screen. Especially for tricky things such as old-style widgets.

Bugs:

- There's an empty widget area at the top of the Block Areas screen which has no title. Probably it's the inactive widget area.
- Widgets added to the inactive widget area in the Widgets screen don't appear at all in Block Areas screen.
- Previewing an old-style widget in the Block Areas screen results in "Block rendered as empty."
- Block Areas screen needs a loading state when you open a widget area. (Or, could we preload everything?)
  - Similarly, Legacy Widget blocks need a loading state.
- _Gutenberg Block_ appears in the dropdown when creating a new Legacy Widget block.
- Need to indicate progress when the Block Area screen's Update button is clicked.
- If you duplicate a Legacy Widget block and then make edits to the duplicate, those changes are saved to both the original and the duplicate widget.

File: phpunit/class-rest-sidebars-controller-test.php
Comment (TimothyBJacobs) [inline phpunit/class-rest-sidebars-controller-test.php:230]: What causes all the empty space? I wonder if it makes sense to `trim` the rendered form?

**Review by TimothyBJacobs — COMMENTED** [review]
Thanks for adding the PHP tests they're looking good! Let's also add coverage for when the user is logged out or has insufficient permissions.

It'd probably also be good to have a test that covers the slashing behavior specifically. Those can get tricky.

**Comment by noisysocks** [comment]
Any thoughts on having a new Gutenberg Block widget vs. using the existing HTML widget?

File: packages/edit-widgets/src/components/widget-areas-block-editor-content/index.js
Comment (jorgefilipecosta) [reply packages/edit-widgets/src/components/widget-areas-block-editor-content/index.js]: Separate editors also bring other challenges like very hard to have a global undo and redo system, more complex to have block navigation that covers all the widget areas, all the a11y to navigate on widget areas would need to be implemented.

File: packages/block-library/src/legacy-widget/edit/dom-manager.js
Comment (jorgefilipecosta) [reply packages/block-library/src/legacy-widget/edit/dom-manager.js:72]: The non-class based widgets render their own form and perform their own save sometimes using tables they created. This data is outside the structure we use to control the widgets and I guess we need a specific save for it, it can be automatically or we can have a save button on the legacy widgets block.

File: packages/edit-widgets/src/components/widget-areas-block-editor-provider/index.js
Comment (jorgefilipecosta) [inline packages/edit-widgets/src/components/widget-areas-block-editor-provider/index.js:62]: What is the reason that makes need to have a post id and use useEntityBlockEditor here? Using it inside the widget area block cause some problem?

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @adamziel thank you for continuing the widgets work, and for all the work on this PR 👍 

Generally, it seems the PR in on a good path, I left some initial comments.

I'm having some problems when testing the branch:
I get the following error:
```
Target container is not a DOM element.
    at http://yah.local/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:25173:28
    at render (http://yah.local/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:25176:7)
    at Module.customizerInitialize (http://yah.local/wp-content/plugins/gutenberg/build/edit-widgets/index.js?ver=2ebbe5d16afdab3466a61f4165697cd3:1914:68)
```

I got this error even when loading the old screen. But on master, I don't have these errors it seems we are now running customizer specific code on both the old and new widgets screen?


I was able to test with success (excluding the errors referred above) the interoperability between the old and new screens I was able to reorder blocks on the old screen :) Nice work accomplishing this!

File: packages/block-library/src/legacy-widget/edit/dom-manager.js
Comment (jorgefilipecosta) [reply packages/block-library/src/legacy-widget/edit/dom-manager.js:72]: Right now the non-class based widgets seem to not be saved.

I tested with this simple widget https://themes.trac.wordpress.org/browser/silesia/1.0.6/include/widgets/feedburner.php
(pasted it on functions.php)


And I verified the values on the form never get saved.

File: lib/widgets.php
Comment (jorgefilipecosta) [inline lib/widgets.php:138]: Right now I'm seeing "Gutenberg block" as an available legacy widget I guess we need to filter it ?

File: lib/widgets-page.php
Comment (noisysocks) [inline lib/widgets-page.php]: We don't need to run _all_ of this function in `widgets.php`, just `wp_enqueue_style( 'wp-block-library' );`

File: lib/class-wp-widget-block.php
Comment (noisysocks) [inline lib/class-wp-widget-block.php:55]: This will sometimes render a `<form>` (e.g. Search block), which is invalid HTML, because we'll then have a `<form>` nested within a `<form>`.

**Review by noisysocks — APPROVED** [review]
I think the direction here is solid and that it's the right way to go. There's lots of remaining work, but we will be here forever addressing it in this already very large PR. Since this is all behind a feature flag, let's merge it now and address the feedback in more manageable follow-up PRs.

Thanks @adamziel! Nice work. I think we can get this feature in for WP 5.6 💪

**Comment by noisysocks** [comment]
I created https://github.com/WordPress/gutenberg/issues/24530 which summarises and tracks the necessary follow-up work here.

File: lib/class-wp-widget-block.php
Comment (adamziel) [reply lib/class-wp-widget-block.php:55]: @noisysocks I don't see a `<form>` anywhere in this file:

https://github.com/WordPress/gutenberg/tree/master/packages/block-library/src/search/edit.js

Where I can see a `<form>` is in the `legacy-widget` block:

https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/legacy-widget/edit/dom-manager.js#L63

I don't think that's a problem though - the widget screen is not wrapped in a `<form>` and so each block is free to render anything it wants.

Comment (noisysocks) [reply lib/class-wp-widget-block.php:55]: Sorry, I wasn't clear here. I'm referring to the legacy `widgets.php` screen. By default every widget is wrapped with a `<form>`.

https://github.com/WordPress/wordpress-develop/blob/master/src/wp-admin/includes/widgets.php#L205

This means that you can sometimes end up with invalid HTML.

1. Navigate to _Widgets (beta)_ and add a Search block to one of the block areas.
2. Navigatie to _Widgets_ (the old screen).
3. View the page source and search for `wp-block-search`.

You'll see that there's a `<form>` inside a `<form>`. Firefox helpfully highlights the invalid HTML.

Screen Shot 2020-08-26 at 11 05 17

**Pull request merged**
